### PR TITLE
feat(cli): change set name is now a constant, and --no-execute will always produce one (even if empty)

### LINKED ISSUE #11075: [cli] Caller should be able to override created changeset name
When invoking `cdk deploy` with the `--no-execute` flag, CDK will create a changeset whose name is completely random. Usually this might be okay, but in automated environments this is unacceptable. As such, it should be possible to provide a name which CDK will use when creating a changeset (e.g. using a `--change-set-name <string>` flag).

### Use Case

Our CI/CD environment runs on top of CodePipeline in combination with CodeBuild (and very extensive use of CDK). Naturally, our initial approach was to simply do a `cdk deploy` inside our CodeBuild containers. However, the stack creation/update itself often takes several minutes with the computing instance just waiting for the stack create to finish. This is not very cost-efficient.
So, to save resources and costs, the CodeBuild containers should not execute the changeset, only create it so that CodePipeline's [CloudFormationExecuteChangeSetAction](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_codepipeline_actions/CloudFormationExecuteChangeSetAction.html) can take over. However, this specific deploy action needs the changeset name which must be known at create time (the [Fn::GetParam](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/continuous-delivery-codepipeline-parameter-override-functions.html#w2ab1c21c20b9) function unfortunately only works for `TemplateOverrides`, tried that already).

### Proposed Solution

I believe the solution is simple: the CLI should accept a `--change-set-name <string>` option, and pass this optional parameter to the [deployStack](https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/lib/api/deploy-stack.ts#L233) function. This one should use the parameter, if provided, or fallback to the existing `uuid.v4()` method. This can also be used in combination with `--execute`, but there it will be significantly less interesting.

### Other

For our use case... we tried going with `cdk synth` and `CreateStackAction` shenanigans, but this leads to many more problems:
- we have lots of assets which will not get uploaded in `cdk synth`, so we'd have to do it all manually or introduce the [cdk-assets](https://github.com/aws/aws-cdk/tree/master/packages/cdk-assets) package.
- CodePipeline uses some other role for the stack than CDK (which is hard to override without diving into CDK internals) and since we have KMS encrypted S3 buckets, this approach led to many hard-to-debug S3 API errors.
- you cannot provide `NotificationArns` with this approach which are also used by us

* [x] :wave: I may be able to implement this feature request

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by swar8080**
closes #11075

Adds two commands to the `deploy` CLI command to make it easier to externally execute change sets when using the `--no-execute` flag:

`--change-set-name`: Optional name of the CloudFormation change set to create, instead of using a random one. An external script or the CodePipeline CloudFormation action can use this name to later deploy the changes.

`--retain-empty-change-set`:  Optionally retain empty change sets instead of deleting them. This is useful for the (requested) CodePipeline use case, since the CodePipeline CloudFormation action always requires a change set, even if it's empty. 

Questions for reviewer:
- Is `--retain-empty-change-set` needed? One workaround for the CodePipeline use case could be for users to write a lambda that generates the required empty change set whenever CDK doesn't generate one. Another idea would be to automatically retain change sets when using `--no-execute` to avoid this extra CLI flag, but this would be a small change in behavior.
- Are the new integration tests overkill? Also should unit tests be added or in-place of the integration tests?

### CODE REVIEW

File: packages/aws-cdk/test/integ/cli/cli.integtest.ts [hunk 1]
```diff
@@ -139,19 +139,58 @@ integTest('nested stack with parameters', withDefaultFixture(async (fixture) =>
   expect(response.StackResources?.length).toEqual(1);
 }));
 
-integTest('deploy without execute', withDefaultFixture(async (fixture) => {
+integTest('deploy without execute a named change set', withDefaultFixture(async (fixture) => {
```
Comment (swar8080) [inline packages/aws-cdk/test/integ/cli/cli.integtest.ts:142]: This is piggy-backing off the existing integration test for the `--no-execute-flag` by also testing the use of a custom name for the change set. I think it's a safe change because it's not changing anything originally being tested, and at code level specifying a name shouldn't really change the previous behavior.

**Review by rix0rrr — CHANGES_REQUESTED** [review]
Honestly -- I would be okay with drastically simplifying the behavior of the CLI. The things you are asking for seem like good default behavior.

* Let's just fix the name of the ChangeSet to be something like `cdk-deploy-changeset`. I don't think there's value in having a UUID as the name (but please correct me if I'm wrong). Is it even possible/feasible to have two independent change sets on the same stack anyway? It seems to me they would have to be serialized in any case, no? (`git rebase`, in essence)
* The behavior of not deleting an empty changeset in case of `--no-execute` seems like good default behavior. It seems weird to me to explicitly ask a tool to create a change set, and then have it not exist after the tool finishes.

The only added complication would be that we would potentially need to delete/replace an existing change set with the name `cdk-deploy-changeset`, but I don't see an issue with that.

Bottom line: let's do away with the flags and make your desired behavior the default. Yes, it's a behavior change, but since we are strengthening postconditions, all tools that potentially exist that currently wrap the CDK CLI and know how to deal with change sets with variable names that might or might not exist, will also successfully deal with a change set with a fixed name that always exists.

**Comment by swar8080** [comment]
> The behavior of not deleting an empty changeset in case of --no-execute seems like good default behavior. It seems weird to me to explicitly ask a tool to create a change set, and then have it not exist after the tool finishes.

Okay sounds good, i'll make that change. The point about the impact on postconditions is interesting to think about for these kind of changes.

>Let's just fix the name of the ChangeSet to be something like cdk-deploy-changeset. I don't think there's value in having a UUID as the name (but please correct me if I'm wrong). Is it even possible/feasible to have two independent change sets on the same stack anyway? It seems to me they would have to be serialized in any case, no? (git rebase, in essence)

The UUID was added because multiple change sets can exist at a time as long as their names are unique, which is currently possible by running  `cdk deploy --no-execute` multiple times. However, only one change set can be executed at a time, which causes all the other change sets to be deleted. 

> The only added complication would be that we would potentially need to delete/replace an existing change set with the name cdk-deploy-changeset, but I don't see an issue with that.

Let me know if the above changes your opinion on deleting the previous change set, since the CloudFormation documentation mentions: _"Creating multiple change sets helps you understand and evaluate how different changes will affect your resources. You can create as many change sets as you need."_

**Comment by rix0rrr** [comment]
> The UUID was added because multiple change sets can exist at a time as long as their names are unique, which is currently possible by running cdk deploy --no-execute multiple times. However, only one change set can be executed at a time, which causes all the other change sets to be deleted.

Right. I believe that's true, but what's the point? Do people actually do this in practice? I'd argue no.

So I'd be fully happy with making sure a changeset with a constant name is created, replacing a previous one with the same name (if it exists).

We *can* still add `--change-set-name` if you feel like you REALLY REALLY want to create 5 different changesets and pick your favorite one to execute... but honestly, let's not and wait until someone asks for it. Feels like that kind of workflow would go against how source controlled infra is supposed to work.  Which version of those 5 would you commit? I'd hope the one that actually gets executed. But why wouldn't you have commited before deploying already?

(In short--I think this workflow is exceedingly rare, and I'd like to be proven wrong before I decide otherwise)

**Review by rix0rrr — DISMISSED** [review]

**Review by rix0rrr — APPROVED** [review]

**Comment by Wenzil** [comment]
> We can still add --change-set-name if you feel like you REALLY REALLY want to create 5 different changesets and pick your favorite one to execute... but honestly, let's not and wait until someone asks for it.

@rix0rrr We have a security use case for this. 

We have the concept of "admin" Change Sets that execute under a very elevated CloudFormation execution role. Admins and normal CDK users both operate on the same stacks using CDK but for security, **normal CDK users shouldn't be able to create or execute the "admin" Change Sets** because of those elevated permissions.

In other words, we have two groups of users working with two mutually exclusive sets of Change Sets on the same stacks.

We want to enforce this using IAM policies in our custom `bootstrap template`. Like so:

```yaml
# AdminDeploymentRole
# ...
- Action: cloudformation:CreateChangeSet
  Resource: '*'
  Effect: Allow
  Condition:
    StringEquals:
      cloudformation:ChangeSetName: "CDK-AdminChangeSet"
      cloudformation:RoleArn:
        Fn::GetAtt: [AdminCloudFormationExecutionRole, Arn] # Elevated role not intended for normal CDK users
- Action:
    - cloudformation:DeleteChangeSet
    - cloudformation:ExecuteChangeSet
  Resource: '*'
  Effect: Allow
  Condition:
    StringEquals:
      cloudformation:ChangeSetName: "CDK-AdminChangeSet"
# ...

# UserDeploymentRole
- Action: cloudformation:CreateChangeSet
  Resource: '*'
  Effect: Allow
  Condition:
    StringEquals:
      cloudformation:ChangeSetName: "CDK-UserChangeSet"
      cloudformation:RoleArn:
        Fn::GetAtt: [UserCloudFormationExecutionRole, Arn]
- Action:
    - cloudformation:DeleteChangeSet
    - cloudformation:ExecuteChangeSet
  Resource: '*'
  Effect: Allow
  Condition:
    StringEquals:
      cloudformation:ChangeSetName: "CDK-UserChangeSet"
# ...
```

**Note**: We are using new-style synthesis. At synth time we are able to control which deployment role to assume (restricted by the trust policy). We use a custom `synthAsAdmin=true` --context flag for that. 

We are just missing the ability to instruct CDK to adequately name the Change Sets (for the above policy to work).

Unless I'm missing an alternative way to achieve the above, this feature is a must have for us.
